### PR TITLE
#[must_use] for functions

### LINKED ISSUE #43302: Tracking issue for RFC 1940: allow `#[must_use]` on functions
This is a tracking issue for the RFC "allow `#[must_use]` on functions" (rust-lang/rfcs#1940).

**Steps:**

- [x] Implement the RFC (cc @rust-lang/compiler -- can anyone write up mentoring instructions?)
- [x] Adjust documentation ([see instructions on forge][doc-guide])
- [x] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

### PR DESCRIPTION
**Pull request by zackmdavis**
This implements [RFC 1940](https://github.com/rust-lang/rfcs/pull/1940).

The RFC and discussion thereof seem to suggest that tagging `PartialEq::eq` and friends as `#[must_use]` would automatically lint for unused comparisons, but it doesn't work out that way (at least the way I've implemented it): unused `.eq` method calls get linted, but not `==` expressions. (The lint operates on the HIR, which sees binary operations as their own thing, even if they ultimately just call `.eq` _&c._.)

What do _you_ think??

Resolves #43302.

### CODE REVIEW

**Comment by zackmdavis** [comment]
Hm, maybe the feature gate is too conservative (the RFC and discussion don't seem to lean that way)?

**Comment by zackmdavis** [comment]
(force-pushed version without feature gate)

**Comment by zackmdavis** [comment]
Wait, sorry, I needed to fix the commit message of what is now 3645b06 (it still mentioned the feature gate).  (The eternal shame of a _factually untrue_ statement blotting the Git history forever is not something I want on my conscience.)

@eddyb reapprove?

File: src/libcore/cmp.rs [hunk 1]
```diff
@@ -625,6 +627,7 @@ pub trait PartialOrd<Rhs: ?Sized = Self>: PartialEq<Rhs> {
     /// let result = std::f64::NAN.partial_cmp(&1.0);
     /// assert_eq!(result, None);
     /// ```
+    #[must_use]
```
Comment (scottmcm) [inline src/libcore/cmp.rs]: This method is the only newly-annotated one that's not an operator.  Intentional?

(I'm by no means against it, and see no situation where it being unused would be good, but I recall the RFC intentionally leaving it off most places for now.)

Comment (zackmdavis) [inline src/libcore/cmp.rs]: Intentional; it felt like "part of the set", even if it's not really an operator.

**Comment by crumblingstatue** [comment]
> unused .eq method calls get linted, but not == expressions

I just want to point out that the [main motivation](https://github.com/rust-lang/rfcs/pull/1812) for reopening this RFC was exactly to lint against unused `==`.

**Comment by alexcrichton** [comment]
Thanks for the PR here @zackmdavis!

I think, though, that the libs team may not have gotten a chance to sign off on the standard-library related changes here. I'll bring this up in the next libs triage to make sure we're on board with it, but if we decide to be more conservative for now, would you be up for backing these out temporarily?

**Pull request merged**